### DISCUSSION TITLE
windows版本的沙箱运行时间有时候是0，这怎么解决？

### DISCUSSION THREAD

**Discussion by xxdcz11** [criyle/go-judge#107] [Q&A] [answered]
您好，我使用的是go-judge的windows amd64版本，但是存在两个问题，第一个是编译和运行c++代码的时候，返回的runtime一直是0，第二个问题是运行c++代码的可执行文件的时候，有一小半情况，返回的time是0，请问这怎么解决？

**Comment by criyle** [comment]
由于编码疏忽，加入 runTime 时未向 windows 接口实现添加此字段实现，会考虑在近期加入。要注意此字段只能提供一个程序运行的观察时间，而不来源于内核统计数据，因此并不准确。

另外，Windows 下时间统计为创建进程之后并不包含进程初始化的开销，并且 time 来自 JobObject 返回的 `TotalUserTime` 并不包含内核时间，而且精度较低 (0.1ms)，因此有可能出现 0 的情况。

**Reply by xxdcz11** [reply]
1.您的意思是如果time小于0.1ms，会直接返回time为0，是这样吗？ 2.如果我使用linux版本的go-judge，runtime和time的问题会解决吗？（我看另外一位使用go-judge的博主，他是linux的go-judge，runtime和time都很正常）

**Comment by xxdcz11** [comment]
1.您的意思是如果time小于0.1ms，会直接返回time为0，是这样吗？       2.如果我使用linux版本的go-judge，runtime和time的问题会解决吗？（我看另外一位使用go-judge的博主，他是linux的go-judge，runtime和time都很正常）

**Reply by criyle** [reply]
根据观察，windows 的计时单位可能是 `2.5ms` 即小于这个时间的运行可能不会被记入。
windows 版本的安全性和 chrome 使用的代码沙箱相当，并没有提供与当前运行环境的强隔离安全性。建议使用 Linux 版本或者在 WSL2 下使用。

**Reply by xxdcz11** [reply]
那请问linux版本是不是没有“低于2.5ms不统计”这个问题？

**Reply by criyle** [reply, marked as answer]
Linux 的时间统计来源于 cgroup 接口提供的精度为 `1ns = 0.000001 ms` 的时间统计，据观察不会出现没有计量的情况。

**Reply by xxdcz11** [reply]
好的，谢谢您解答我的疑问。祝您学习工作顺利！

**Comment by xxdcz11** [comment]
使用体验反馈：在WSL2下安装Ubuntu，然后安装go-judge，使用之后发现，runtime和time问题已经解决，而且使用Linux版本的沙箱，相比于Windows版本的沙箱，速度也是提高了很多很多。

**Comment by xxdcz11** [comment]
使用体验反馈：在WSL2下安装Ubuntu，然后安装go-judge，使用之后发现，runtime和time问题已经解决，而且使用Linux版本的沙箱，相比于Windows版本的沙箱，速度也是提高了很多很多。
